### ISSUE TITLE
Kubelet doesn't retry if failed to get ECR auth token

### ISSUE THREAD

**Issue by andrewsykim** [kubernetes/kubernetes#37962]
---
AWS has a rate limit of 1 ECR auth token per second per account. For kubernetes cluster with many nodes and using ECR as its primary registry its not unlikely for the kubelet to get throttled. 
```
aws_credentials.go:199] while requesting ECR authorization token ThrottlingException: Rate exceeded
```
After digging through some of the code in the [aws_credentials.go](https://github.com/kubernetes/kubernetes/blob/master/pkg/credentialprovider/aws/aws_credentials.go)  (and please correct me if I'm wrong) it seems like the kubelet does not implement any retry logic for getting the ECR token which leaves that worker node completely useless until the expiration time for the token (11 hours and 55 minutes). The only way to trigger another fetch is to restart the kublet. The kubelet should either block until the token is acquired OR execute some retry loop in a separate goroutine. 

**Kubernetes version** (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"4", GitVersion:"v1.4.4", GitCommit:"3b417cc4ccd1b8f38ff9ec96bb50a81ca0ea9d56", GitTreeState:"clean", BuildDate:"2016-10-21T02:48:38Z", GoVersion:"go1.7.1", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"4", GitVersion:"v1.4.4", GitCommit:"3b417cc4ccd1b8f38ff9ec96bb50a81ca0ea9d56", GitTreeState:"clean", BuildDate:"2016-10-21T02:42:39Z", GoVersion:"go1.6.3", Compiler:"gc", Platform:"linux/amd64"}
```

**Environment**:
- **Cloud provider or hardware configuration**: AWS
- **OS** (e.g. from /etc/os-release): 
```
NAME=CoreOS
ID=coreos
VERSION=1185.2.0
VERSION_ID=1185.2.0
BUILD_ID=2016-10-20-2326
PRETTY_NAME="CoreOS 1185.2.0 (MoreOS)"
ANSI_COLOR="1;32"
HOME_URL="https://coreos.com/"
BUG_REPORT_URL="https://github.com/coreos/bugs/issues"
```
- **Kernel** (e.g. `uname -a`):
```
Linux 4.7.3-coreos-r2 #1 SMP Thu Oct 20 23:18:54 UTC 2016 x86_64 Intel(R) Xeon(R) CPU E5-2680 v2 @ 2.80GHz GenuineIntel GNU/Linux
```

**Comment by justinsb** [comment]
Oh wow - overlooked this issue and PR - sorry!  And cc @therc who wrote the original code and knows it better than I do.

I have a few thoughts:

1) Normally the aws-sdk-go does a reasonable job of retries.  So the error message is surprising.  We do have a custom retry policy in the "main" cloudprovider, which we should use.  But it actually is primarily for logging, and actually to insert a process-wide backoff (IIRC).  The process-wide backoff is needed because otherwise rate limiting is done per call, and concurrent calls can easily exhaust limits.  That might be what is happening here.
2) I do wonder what the default retry policy is for ECR.  It may also have changed recently, because we updated aws-sdk-go.
3) Separately though, a 5 minute outage of ECR (as a thought experiment) shouldn't take us offline for 12 hours.  I propose we need should not treat an error response as identical to the happy path.

In other words, I wonder if we should fix this as well:
https://github.com/kubernetes/kubernetes/blob/8fd414537b5143ab039cb910590237cabf4af783/pkg/credentialprovider/provider.go#L120-L121

I don't mind the aggressive retries - though I worry that we are knowingly going to the max for 20 seconds, and this limit is AWS _account_ wide I believe.  But there will always be problems, right, and so I feel the underlying problem is that it takes us 12 hours to recover, instead of a few minutes...

PS thank you so much for finding this - I think this explains so much.  And sorry that I didn't find it sooner in my email :-(

**Comment by andrewsykim** [comment]
@justinsb no worries! Thanks for responding. 

Based off a quick look at the aws credential code (https://github.com/kubernetes/kubernetes/blob/master/pkg/credentialprovider/aws/aws_credentials.go), it seems like its decoupled from the main aws code we use in the controller manager. That might explain why there's no retry logic. Maybe a good place to start could be importing the main AWS code into the credentials provider?

One alternative that I wanted to experiment with is: https://github.com/upmc-enterprises/registry-creds which inserts ecr auth token into `imagePullSecrets` in the default service account. Until we fix this in kubernetes core we may want to promote this alternative. It would also be much easier to implement retry logic in the addon rather then in the kubelet. 

What do you think?

**Comment by bigkraig** [comment]
@justinsb @andrewsykim We get the throttling message all of the time on our accounts when using https://github.com/rlister/ecr-login, so I am not sure if anything is built into the SDK. Would really like to see this one fixed though, our kubelets hit it all of the time.

**Comment by bigkraig** [comment]
@justinsb @andrewsykim if the retry logic we are depending on is in aws-sdk-go then I don't think it's being used, because we hit the rate limit all of the time when using https://github.com/rlister/ecr-login/blob/master/ecr-login.go. If #37963 is a candidate for a pull, I can fix it up.

**Comment by andrewsykim** [comment]
@bigkraig I haven't tried this yet, but I recommend https://github.com/upmc-enterprises/registry-creds

**Comment by andrewsykim** [comment]
Feel free to fix up that PR btw :)
